src/vm/dllimport.cpp

```diff
@@ -4139,10 +4363,150 @@ static void CreateNDirectStubWorker(StubState*         pss,
     pss->FinishEmit(pMD);
 }
 
+static CorNativeLinkType GetLinkTypeOfMethodTable(MethodTable* pMT)
+{
+    CorNativeLinkType nltType;
+
+    IMDInternalImport* pInternalImport = pMT->GetModule()->GetMDImport();
+
+    DWORD clFlags;
+    if (FAILED(pInternalImport->GetTypeDefProps(pMT->GetTypeDefRid(), &clFlags, NULL)))
+    {
+        UNREACHABLE_MSG("Structs that are generating interop marshalling stubs have already been verified to have valid metadata");
+    }
+
+    if (IsTdAnsiClass(clFlags))
+    {
+        nltType = nltAnsi;
+    }
+    else if (IsTdUnicodeClass(clFlags))
+    {
+        nltType = nltUnicode;
+    }
+    else if (IsTdAutoClass(clFlags))
+    {
+#ifdef PLATFORM_WINDOWS
+        nltType = nltUnicode;
+#else
+        nltType = nltAnsi; // We don't have a utf8 charset in metadata yet, but ANSI == UTF-8 off-Windows
+#endif
+    }
+    else
+    {
+        UNREACHABLE_MSG("Structs that are generating interop marshalling stubs have already been verified to have valid metadata");
+    }
+
+    return nltType;
+}
+
+static void CreateStructStub(ILStubState* pss,
+    StubSigDesc* pSigDesc,
+    MethodTable* pMT,
+    DWORD dwStubFlags,
+    MethodDesc* pMD)
+{
+    CONTRACTL
+    {
+        STANDARD_VM_CHECK;
+        PRECONDITION(CheckPointer(pss));
+        PRECONDITION(CheckPointer(pSigDesc));
+        PRECONDITION(CheckPointer(pMD, NULL_OK));
+        PRECONDITION(!pMD || pMD->IsILStub());
+        PRECONDITION(SF_IsStructMarshalStub(dwStubFlags));
+    }
+    CONTRACTL_END;
+
+
+    SF_ConsistencyCheck(dwStubFlags);
+
+#ifdef _DEBUG
+    if (g_pConfig->ShouldBreakOnInteropStubSetup(pSigDesc->m_pDebugName))
+        CONSISTENCY_CHECK_MSGF(false, ("BreakOnInteropStubSetup: '%s' ", pSigDesc->m_pDebugName));
+#endif // _DEBUG
+
+    Module* pModule = pSigDesc->m_pModule;
+
+
+    pss->SetLastError(false);
+
+    pss->BeginEmit(dwStubFlags);
+
+    // Marshal the fields
+    MarshalInfo::MarshalScenario ms = MarshalInfo::MARSHAL_SCENARIO_FIELD;
+#ifdef FEATURE_COMINTEROP
+    if (pMT->IsProjectedFromWinRT())
+    {
+        ms = MarshalInfo::MARSHAL_SCENARIO_WINRT_FIELD;
+    }
+#endif // FEATURE_COMINTEROP
+
+    int numFields = pMT->GetNumInstanceFields();
+    // Build up marshaling information for each of the method's parameters
+    SIZE_T cbFieldMarshalInfo;
+    if (!ClrSafeInt<SIZE_T>::multiply(sizeof(MarshalInfo), numFields, cbFieldMarshalInfo))
+    {
+        COMPlusThrowHR(COR_E_OVERFLOW);
+    }
+
+    CorNativeLinkType nlType = GetLinkTypeOfMethodTable(pMT);
+    NativeFieldDescriptor* pFieldDescriptors = pMT->GetLayoutInfo()->GetNativeFieldDescriptors();
+
+
+    for (int i = 0; i < numFields; ++i)
```

Comment:
Iterator style.

Reply:
We use the value of `i` a few times below, so I think this loop is better left as an indexing loop.